### PR TITLE
Widen importlib-* backport deps, add python_version limits.

### PR DESCRIPTION
**Pull request by mjpieters**
Later versions either add small features not used in Airflow, packaging updates and bug fixes or perf improvements. Moreover, newer Python releases already include the functionality needed.

It's generally a better idea to be permissive in dependency versions to prevent
accidentally limiting what other packages can be used with Airflow.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: setup.cfg
Comment (ashb) [inline setup.cfg:103]: Do we not need to up the min requirements for argcomplete too?

**Review by ashb — COMMENTED** [review]
I think there's a use of importlib_resources that should be correct too.

File: setup.cfg
Comment (mjpieters) [reply setup.cfg:103]: Not really. `argcomplete~=1.10` is the same thing as `argcomplete>=1.10,<2`, and so `1.22` is included in that range.

Comment (mjpieters) [reply setup.cfg:103]: Put differently: that'd be a separate task if someone wanted to update this explicitly. But unless you depend on the bugfixes in later releases, I'd leave this be. No need to make the dependency link tighter.

**Comment by mjpieters** [comment]
> I think there's a use of importlib_resources that should be correct too.

The single import is already guarded.

File: setup.cfg
Comment (ashb) [reply setup.cfg:103]: Historical we've updated the intermediary rep, otherwise I thought:

Pip would leave argparse at 1.10, which has a dep of importlib_metadata, but then this new dep would install a newer importlib_metadata, leading to incompatible version errors.

Particularly for "older" pips. Am I misremembering?

Comment (mjpieters) [reply setup.cfg:103]: When you use `pip install -U Apache-airflow` to upgrade airflow, a newer `argcomplete` would be installed, automatically.

In my experience it takes installing additional packages with incompatible dependencies for things to break; newer `pip` versions warn when this happens however. Running `pip check` lets you list the problem packages.

Comment (potiuk) [reply setup.cfg:103]: The problem might be that in our process (in master merges) we automatically do 'pip install .[all] --upgrade --upgrade-strategy eager` to automatically find set of latest constraints that are matching setup.py + setup.cfg limitations, Then we run pip check and then we run all tests and if it all passes we automatically commit the new constraints.

As far as I remember without ~1.* for importlib_resources in install_requires caused a problem as pip first tries to install the highest version matching in `install_requires` only - without looking at extras. Once it installs this version, it will only then try to do eager update on all other extras and one of our dependencies already has importib_resources ~= 1.4. Eager upgrade will never downgrade the packages installed by install_requires, so this results in version conflict.

**Review by potiuk — CHANGES_REQUESTED** [review]

File: setup.cfg
Comment (mjpieters) [reply setup.cfg:103]: Thinking about this some more: we did loosen the restrictions on what versions are acceptable but `importlib_metadata` *is* a new dependency relative to 2.0.0b3. If you then use `pip install apache-airflow>2.0.0b3` (say) *without* `-U` then I think you could indeed end up with an importlib_metadata that is incompatible with an existing argcomplete version if you only have argparse 1.10.

Want me to up the version on that package too, then?

Comment (potiuk) [reply setup.cfg:103]: In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory? 

With those labels, the PR will automatically try to do what master merge does - it will try to figure out latest constraints with eager upgrade strategy and it will do it for all python versions, so let's see if it works.

Comment (potiuk) [reply setup.cfg:103]: BTW. We do not recommend running unconstrained upgrade of airflow - this has caused us terrible problems in the past. That's why we have this constraint mechanism and with the recent changes (I am also fixing that for 1.10.14) we take care about making sure that the set of constraints that we recommend (in INSTALL.txt) are non-conflicting and tested (and we update the constraints-1-10 branch and constrains-master automatically to include latest releases of dependencies (all tests muss pass and pip check must show no-conflicts for those as of Friday)

**Comment by ashb** [comment]
> > I think there's a use of importlib_resources that should be correct too.
> 
> The [single import](https://github.com/apache/airflow/blob/2037303eef93fd36ab13746b045d1c1fee6aa143/airflow/providers_manager.py#L32-L36) I found is already guarded. Or were you thinking of something else?

Oh no that was it. Thought it wasn't guarded for some reason

File: setup.cfg
Comment (potiuk) [reply setup.cfg:103]: Confirmed: @mjpieters @ashb 

importlib_resources~=1.4 is a requirement of cnf-lint which is requirement of moto 1.3.16 which we have. So anything above 1 will cause version conflict. mot 1.3.16 is the latest available version, so I do not think we can do anything about it.

```
moto==1.3.16
  - aws-xray-sdk [required: >=0.93,!=0.96, installed: 2.6.0]
    - botocore [required: >=1.11.3, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - future [required: Any, installed: 0.18.2]
    - jsonpickle [required: Any, installed: 1.4.1]
      - importlib-metadata [required: Any, installed: 1.7.0]
        - zipp [required: >=0.5, installed: 3.4.0]
    - wrapt [required: Any, installed: 1.12.1]
  - boto [required: >=2.36.0, installed: 2.49.0]
  - boto3 [required: >=1.9.201, installed: 1.15.18]
    - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
      - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
        - six [required: >=1.5, installed: 1.15.0]
      - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
      - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
          - six [required: >=1.5, installed: 1.15.0]
        - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - botocore [required: >=1.12.201, installed: 1.18.18]
    - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
    - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
      - six [required: >=1.5, installed: 1.15.0]
    - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
  - cfn-lint [required: >=0.4.0, installed: 0.42.0]
    - aws-sam-translator [required: >=1.25.0, installed: 1.31.0]
      - boto3 [required: ~=1.5, installed: 1.15.18]
        - botocore [required: >=1.18.18,<1.19.0, installed: 1.18.18]
          - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
          - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
            - six [required: >=1.5, installed: 1.15.0]
          - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
        - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
        - s3transfer [required: >=0.3.0,<0.4.0, installed: 0.3.3]
          - botocore [required: >=1.12.36,<2.0a.0, installed: 1.18.18]
            - jmespath [required: >=0.7.1,<1.0.0, installed: 0.10.0]
            - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.1]
              - six [required: >=1.5, installed: 1.15.0]
            - urllib3 [required: >=1.20,<1.26, installed: 1.25.11]
      - jsonschema [required: ~=3.2, installed: 3.2.0]
        - attrs [required: >=17.4.0, installed: 20.3.0]
        - importlib-metadata [required: Any, installed: 1.7.0]
          - zipp [required: >=0.5, installed: 3.4.0]
        - pyrsistent [required: >=0.14.0, installed: 0.17.3]
        - setuptools [required: Any, installed: 50.3.2]
        - six [required: >=1.11.0, installed: 1.15.0]
      - six [required: ~=1.15, installed: 1.15.0]
    - importlib-resources [required: ~=1.4, installed: 1.5.0]
```

Comment (ashb) [reply setup.cfg:103]: Yet again pip's lack of a fully featured dependency solver bites us

**Comment by ashb** [comment]
Opened https://github.com/aws-cloudformation/cfn-python-lint/pull/1808 -- not sure how quickly that will get merged and released.

File: setup.cfg
Comment (mjpieters) [reply setup.cfg:103]: > In order to test it out, I've added the labels "upgrade to latest dependencies" and "full tests needed". Can you please @mjpieters amend and push --force-with-lease to test my theory?

I rebased and pushed before I noticed that a bot had removed the tags again.

I've pushed a change now that ensures we only allow `importlib-resources` 1.x (starting with `1.4`)

Comment (potiuk) [reply setup.cfg:103]: yeah - full tests needed only works after the approval status is "Green" (whch is actually good - we can see if it wiorks for 3.6 first and once it does, we approve and the label will be added and we rerun it again with full matrix of tests.

**Comment by potiuk** [comment]
`No broken requirements found.` for 3.6 :)

File: setup.cfg
Comment (potiuk) [reply setup.cfg:103]: Let's see if it works. I will approve it tentatively, and please rebase @mjpieters - let's see if it will work.

**Review by potiuk — APPROVED** [review]
Approved to get the label working. But let's not merge it until IP check succeeds for all python versions.

**Comment by potiuk** [comment]
Can you please rebase @jpieters ?

**Comment by mjpieters** [comment]
@potiuk I don't know if the Google provider package import failure is related?

```
Traceback (most recent call last):
  File "/opt/airflow/dev/import_all_classes.py", line 71, in import_all_classes
    _module = importlib.import_module(modinfo.name)
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/airflow/airflow/providers/google/cloud/sensors/pubsub.py", line 25, in <module>
    from airflow.providers.google.cloud.hooks.pubsub import PubSubHook
  File "/opt/airflow/airflow/providers/google/cloud/hooks/pubsub.py", line 29, in <module>
    from google.cloud.pubsub_v1.types import (
ImportError: cannot import name 'RetryPolicy'
```

Would that be caused by an ancient version having been pulled in?

**Comment by mjpieters** [comment]
Nope, it's that Google updated that package and moved the class, see https://github.com/googleapis/python-pubsub/commit/6b9eec81ccee81ab93646eaf7652139fc218ed36

**Comment by potiuk** [comment]
> Nope, it's that Google updated that package and moved the class, see [googleapis/python-pubsub@6b9eec8](https://github.com/googleapis/python-pubsub/commit/6b9eec81ccee81ab93646eaf7652139fc218ed36)

I think we had <2.0.0 limit on all Google Python libraries. They released all libs in their 2.0 version - those are not backwards-compatibe and we have (or I thought so) limited those libs to <2.0 for all google lib (until we rewrite the operators)

**Comment by potiuk** [comment]
> > Nope, it's that Google updated that package and moved the class, see [googleapis/python-pubsub@6b9eec8](https://github.com/googleapis/python-pubsub/commit/6b9eec81ccee81ab93646eaf7652139fc218ed36)
> 
> I think we had <2.0.0 limit on all Google Python libraries. They released all libs in their 2.0 version - those are not backwards-compatibe and we have (or I thought so) limited those libs to <2.0 for all google lib (until we rewrite the operators): 

'google-cloud-pubsub>=1.0.0,<2.0.0',

**Comment by potiuk** [comment]
I just downloaded the image built for your PR and inspecting it: 

The good thing is that we are storing all the images from cI in the registry and we can always reproduce things locally:

`./breeze --github-image-id 392290560 --python 3.6` in this case drops you inside the very image that was built in your PR

**Comment by potiuk** [comment]
pip freeze  |grep pubsub
google-cloud-pubsub==1.7.0

**Comment by potiuk** [comment]
That works: 

>>> from google.cloud.pubsub_v1.types import RetryPolicy
>>>

**Comment by potiuk** [comment]
>>> from airflow.providers.google.cloud.hooks.pubsub import PubSubHook
>>>

**Comment by potiuk** [comment]
Interesting thing that in the same image all those work too: 

>>> import importlib
>>> importlib.import_module('airflow.providers.google.cloud.example_dags.example_pubsub')
<module 'airflow.providers.google.cloud.example_dags.example_pubsub' from '/opt/airflow/airflow/providers/google/cloud/example_dags/example_pubsub.py'>
>>> exit 
Use exit() or Ctrl-D (i.e. EOF) to exit
>>> exit()
root@02a0c7718f03:/opt/airflow# python
Python 3.6.12 (default, Oct 13 2020, 21:45:01) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import importlib
>>> importlib.import_module('airflow.providers.google.cloud.example_dags.example_pubsub')
<module 'airflow.providers.google.cloud.example_dags.example_pubsub' from '/opt/airflow/airflow/providers/google/cloud/example_dags/example_pubsub.py'>
>>> from airflow.providers.google.cloud.operators.pubsub import (
...     PubSubCreateSubscriptionOperator,
...     PubSubCreateTopicOperator,
...     PubSubDeleteSubscriptionOperator,
...     PubSubDeleteTopicOperator,
...     PubSubPublishMessageOperator,
...     PubSubPullOperator,
... )
>>>

**Comment by potiuk** [comment]
OK. I reproduced it : docker run -it --entrypoint /usr/local/bin/dumb-init docker.pkg.github.com/apache/airflow/master-python3.6-ci:392290560 -- /opt/airflow/scripts/in_container/run_prepare_provider_readme.sh

**Comment by potiuk** [comment]
The problem is this line: "Installing remaining packages from 'all' extras"

We are running all remaining packages from "devel_all" extras: `pip install -e ".[devel_all]"`. which installs "apache-beam"  -but apparently this causes a host of downgrades: 

```
  Attempting uninstall: cachetools
    Found existing installation: cachetools 4.1.1
    Uninstalling cachetools-4.1.1:
      Successfully uninstalled cachetools-4.1.1
  Attempting uninstall: httplib2
    Found existing installation: httplib2 0.18.1
    Uninstalling httplib2-0.18.1:
      Successfully uninstalled httplib2-0.18.1
  Attempting uninstall: google-resumable-media
    Found existing installation: google-resumable-media 1.1.0
    Uninstalling google-resumable-media-1.1.0:
      Successfully uninstalled google-resumable-media-1.1.0
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 2.0.0
    Uninstalling pyarrow-2.0.0:
      Successfully uninstalled pyarrow-2.0.0
  Attempting uninstall: mock
    Found existing installation: mock 4.0.2
    Uninstalling mock-4.0.2:
      Successfully uninstalled mock-4.0.2
  Attempting uninstall: google-cloud-bigquery-storage
    Found existing installation: google-cloud-bigquery-storage 2.1.0
    Uninstalling google-cloud-bigquery-storage-2.1.0:
      Successfully uninstalled google-cloud-bigquery-storage-2.1.0
  Attempting uninstall: google-cloud-bigquery
    Found existing installation: google-cloud-bigquery 2.4.0
    Uninstalling google-cloud-bigquery-2.4.0:
      Successfully uninstalled google-cloud-bigquery-2.4.0
  Attempting uninstall: fastavro
    Found existing installation: fastavro 1.2.0
    Uninstalling fastavro-1.2.0:
      Successfully uninstalled fastavro-1.2.0
  Attempting uninstall: dill
    Found existing installation: dill 0.3.3
    Uninstalling dill-0.3.3:
      Successfully uninstalled dill-0.3.3
  Attempting uninstall: google-cloud-vision
    Found existing installation: google-cloud-vision 1.0.0
    Uninstalling google-cloud-vision-1.0.0:
      Successfully uninstalled google-cloud-vision-1.0.0
  Attempting uninstall: google-cloud-videointelligence
    Found existing installation: google-cloud-videointelligence 1.16.1
    Uninstalling google-cloud-videointelligence-1.16.1:
      Successfully uninstalled google-cloud-videointelligence-1.16.1
  Attempting uninstall: google-cloud-spanner
    Found existing installation: google-cloud-spanner 1.19.1
    Uninstalling google-cloud-spanner-1.19.1:
      Successfully uninstalled google-cloud-spanner-1.19.1
  Attempting uninstall: google-cloud-pubsub
    Found existing installation: google-cloud-pubsub 1.7.0
    Uninstalling google-cloud-pubsub-1.7.0:
      Successfully uninstalled google-cloud-pubsub-1.7.0
  Attempting uninstall: google-cloud-dlp
    Found existing installation: google-cloud-dlp 1.0.0
    Uninstalling google-cloud-dlp-1.0.0:
      Successfully uninstalled google-cloud-dlp-1.0.0
  Attempting uninstall: google-cloud-bigtable
    Found existing installation: google-cloud-bigtable 1.6.0
    Uninstalling google-cloud-bigtable-1.6.0:
      Successfully uninstalled google-cloud-bigtable-1.6.0
  Attempting uninstall: google-cloud-storage
    Found existing installation: google-cloud-storage 1.33.0
    Uninstalling google-cloud-storage-1.33.0:
      Successfully uninstalled google-cloud-storage-1.33.0
  Attempting uninstall: apache-airflow
    Found existing installation: apache-airflow 2.0.0b3
    Uninstalling apache-airflow-2.0.0b3:
      Successfully uninstalled apache-airflow-2.0.0b3
  Running setup.py develop for apache-airflow
Successfully installed apache-airflow apache-beam-2.23.0 avro-python3-1.9.2.1 cachetools-3.1.1 crcmod-1.7 dill-0.3.1.1 fastavro-0.23.6 fasteners-0.15 google-apitools-0.5.31 google-cloud-bigquery-1.24.0 google-cloud-bigquery-storage-1.1.0 google-cloud-bigtable-1.0.0 google-cloud-datastore-1.7.4 google-cloud-dlp-0.13.0 google-cloud-pubsub-1.0.2 google-cloud-spanner-1.13.0 google-cloud-storage-1.29.0 google-cloud-videointelligence-1.13.0 google-cloud-vision-0.42.0 google-resumable-media-0.5.1 httplib2-0.17.4 mock-2.0.0 monotonic-1.5 oauth2client-3.0.0 pyarrow-0.17.1 pydot-1.4.1
````

Including `google-cloud-pubsub-1.0.2`

Apache-Beam is the only excluded package for CI image as it used to be known by really bad requirements, forcing many packages to downgrade.

`
# Those packages are excluded because they break tests (downgrading mock) and they are
# not needed to run our test suite.
PACKAGES_EXCLUDED_FOR_CI = [
    'apache-beam',
]
`

We are however installing it for provider package check to be able to import DataProc.

I was actually going to take a look at apache-beam for 2.0 now. I discussed it today with @mik-laj that the recent 'apache-beam' has better requirements and we should be able to install it without breaking too much - obviously not that version.

**Comment by potiuk** [comment]
I will take a look at not excluding apache-beam tomorrow.

**Comment by potiuk** [comment]
Also what might help is this PR which I have to finish: https://github.com/apache/airflow/pull/12693 

We are not using constraints in this install, which is the reason why so many packages got downgraded.

I will see to it tomorrow. Stay tuned.

**Comment by potiuk** [comment]
BTW. It is super easy to reproduce it all thanks to the image storage in registry. This is exactly why I implemented this: https://github.com/apache/airflow/pull/12718 - it will tell you exactly what command you should run to be dropped to the same image that is used in CI and you can run exactly what was run in CI and get perfect reproduction capabilities (I did it now and once we merge this PR, you will get those instructions how to get there:

See this: https://github.com/apache/airflow/pull/12718/checks?check_run_id=1475794405#step:4:564.

**Comment by potiuk** [comment]
Added https://github.com/apache/airflow/pull/12703 to show output of this "devel_all" installation in logs

**Comment by ashb** [comment]
The change to cfn-lint has merged, from previous history it looks like they release every week or two.

**Pull request closed without merging**
